eos/effects/adaptivearmorhardener.py

```diff
@@ -2,17 +2,207 @@
+
+            # Sort the tuple by which resist took the most damage
+            damagePattern_tuple = sorted(damagePattern_tuple, key=operator.itemgetter(3,0))
+
+            if damagePattern.emAmount == damagePattern.thermalAmount == damagePattern.kineticAmount == damagePattern.explosiveAmount:
+                # If damage pattern is even across the board, we "reset" back to original resist values.
+                logger.debug("Setting adaptivearmorhardener resists to uniform profile.")
+                # Do nothing, because the RAH gets recalculated, so just don't change it
+                runLoop = 0
+            elif damagePattern_tuple[0][2] == damagePattern_tuple[1][2] == damagePattern_tuple[2][2] == 0:
+                # If damage pattern is a single source, we set all resists to one damage profile.
+                logger.debug("Setting adaptivearmorhardener resists to single damage profile.")
+
+                #Do this dynamically just in case CCP mucks with the values.
+                vampDmg=0
+                for i in [0,1,2]:
+                    attr = "armor%sDamageResonance" % damagePattern_tuple[i][0].capitalize()
+                    vampDmg = vampDmg+1-damagePattern_tuple[i][4]
+                    module.increaseItemAttr(attr, (1-damagePattern_tuple[i][4]))
+                    damagePattern_tuple[i][4] = 1
+
+                attr = "armor%sDamageResonance" % damagePattern_tuple[3][0].capitalize()
+                damagePattern_tuple[3][4]=damagePattern_tuple[3][4]-vampDmg
+                module.increaseItemAttr(attr, 0-vampDmg)
+                runLoop = 0
+            else:
+                #logger.debug("Multiple damage type profile.")
+                if damagePattern_tuple[1][4] == 1 == damagePattern_tuple[0][4]:
+                    logger.debug("We've run out of resists to steal. Breaking out of RAH cycle.")
+                    break
+                elif (countPasses >= 100) or (countPassesDos >= 5):
+                    logger.debug("Looped a total of %f times, and %f times after resistance shift amount was reduced. Most likely the RAH is cycling between two different profiles and is in an infinite loop. Breaking out of RAH cycle.", countPasses, countPassesDos)
+                    # If we hit this break point and have over 100 passes, something has gone horribly wrong.
+                    # Most likely we'll hit this after we've reduced the resist shifting down to .01 and looped 5 times.
+                    break
+                elif (damagePattern_tuple[0][4] == 1) or (countPasses >= 15):
+                    # Most likely the RAH is cycling between two different profiles and is in an infinite loop.
+                    # Reduce the amount of resists we shift each loop, to try and stabilize on a more average profile.
+                    if resistanceShiftAmount > .01:
+                        resistanceShiftAmount = resistanceShiftAmount - .01
```

Comment:
In theory that should work something like this: http://imgur.com/le3bwsg
If it worked that way it would be a great solution. The problem is that most of the time when the RAH enters a loop it's bouncing around in such a way that it can't converge on the average point. Reducing the shift amount between cycles tends to result in the RAH getting stuck at a point completely unrelated to the average of the loop cycles, or even any of the cycles themselves. I imagine it kind of like this: http://imgur.com/6VxrL3Z